File: Days.33/C++/4.BST.cpp
// Binary Search Tree

#include<iostream>
using namespace std;

class Node
{
    public:
    int val;
    Node *left;
    Node *right;
    Node(int val)
    {
        this->val=val;
        this->left=this->right=nullptr;
    }
};

class BST
{
    public:
    Node *root;
    BST()
    {
        root=nullptr;
    }
    Node *Insert(Node *root,int val)
    {
        if (root==nullptr)
        {
            return new Node(val);
        }
        if (root->val==val)
        {
            return root;
        }
        if (root->val>val)
        {
            root->left=Insert(root->left,val);
        }
        else
        {
            root->right=Insert(root->right,val);
        }
        return root;
    }

    bool Search(Node *root,int val)
    {
        if(root==nullptr)
        {
            return false;
        }
        if(root->val==val)
        {
            return true;
        }
        if (root->val>val)
        {
            return Search(root->left,val);
        }
        else{
            return Search(root->right,val);
        }
    }

    Node* Delete(Node *root,int val)
    {
        if (root==nullptr)
        {
            return nullptr;
        }
        if(root->val==val)
        {
            if(root->left!=nullptr and root->right!=nullptr)
            {
                Node *temp=root->left;
                while (temp->right!=nullptr)
                {
                    temp=temp->right;
                }
                root->val=temp->val;
                root->left=Delete(root->left,temp->val);
                return root;
            }
            if(root->left!=nullptr)
            {
                return root->left;
            }
            else{
                return root->right;
            }
        }
        if(root->val>val)
        {
            root->left=Delete(root->left,val);
        }
        else{
            root->right=Delete(root->right,val);
        }
        return root;
    }

    void Inorder(Node *root)
    {
        if (root==nullptr)
        {
            return;
        }
        Inorder(root->left);
        cout<<root->val<<" ";
        Inorder(root->right);
    }
};

int main()
{
    BST *bst=new BST();
    int val,choice;
    while (true)
    {
        cout<<"1.Insert\n";
        cout<<"2.Delete\n";
        cout<<"3.Search\n";
        cout<<"4.Inorder Traversal\n";
        cout<<"5.Exit\n";
        cout<<"Enter Your Choice:\t";
        cin>>choice;
        if (choice==1)
        {
            cout<<"Enter a Value:\t";
            cin>>val;
            bst->root=bst->Insert(bst->root,val);
        }
        else if(choice==2)
        {
            cout<<"Enter Value to be Deleted:\t";
            cin>>val;
            if(bst->Search(bst->root,val))
            {
                bst->root=bst->Delete(bst->root,val);
                cout<<"Deleted Successfully\n";
            }
            else{
                cout<<"Not Present in BST\n";
            }
        }
        else if(choice==3)
        {
            cout<<"Enter value for Searching:\t";
            cin>>val;
            if (bst->Search(bst->root,val))
            {
                cout<<"Present in BST\n";
            }
            else
            {
                cout<<"Not Present in BST\n";
            }
        }
        else if(choice==4)
        {
            bst->Inorder(bst->root);
            cout<<"\n";
        }
        else if(choice==5)
        {
            break;
        }
        else{
            cout<<"Invalid Choice\n";
        }
    }
    return 0;
}

File: Days.33/Python/4.BST.py

class Node:
    def __init__(self,val):
        self.val=val
        self.left=None
        self.right=None

class BST:
    def __init__(self):
        self.root=None
    def Insert(self,root,val):
        if root==None:
            return Node(val)
        if root.val==val:
            return root
        elif root.val>val:
            root.left=self.Insert(root.left,val)
        else:
            root.right=self.Insert(root.right,val)
        return root
    
    def Search(self,root,val):
        if root==None:
            return False
        if root.val==val:
            return True
        if root.val>val:
            return self.Search(root.left,val)
        else:
            return self.Search(root.right,val)

    def Delete(self,root,val):
        if root==None:
            return None
        if root.val==val:
            if root.left==None and root.right==None:
                return None
            elif root.left!=None and root.right!=None:
                temp=root.left
                while temp.right!=None:
                    temp=temp.right
                root.val=temp.val
                root.left= self.Delete(root.left,temp.val)
                return root
            elif root.left!=None:
                return root.left
            else:
                return root.right
        if root.val>val:
            root.left= self.Delete(root.left,val)
        else:
            root.right=self.Delete(root.right,val)
        return root

    def Inorder(self,root):
        if root==None:
            return
        self.Inorder(root.left)
        print(root.val,end=" ")
        self.Inorder(root.right)


def main():
    bst=BST()
    while True:
        print('1.Insert')
        print('2.Delete')
        print('3.Search')
        print('4.Inorder Traversal')
        print('5.Exit')
        choice=int(input('Enter Your Choice: '))
        if choice==1:
            val=int(input('Enter a Value: '))
            bst.root=bst.Insert(bst.root,val)
        elif choice==2:
            val=int(input('Enter Value to be Deleted: '))
            flag=False
            if bst.Search(bst.root,val):
                flag=True
                bst.root=bst.Delete(bst.root,val)
            if flag:
                print('Deleted Successfully')
            else:
                print('Value is Not Present in BST')
        elif choice==3:
            val=int(input('Enter value for searching: '))
            if bst.Search(bst.root,val):
                print('Present in BST')
            else:
                print('Not Present in BST')
        elif choice==4:
            bst.Inorder(bst.root)
            print()
        elif choice==5:
            break
        else:
            print('Invalid Choice')



main()

